File: 3310.remove-methods-from-project.cpp
#
# @lc app=leetcode id=3310 lang=cpp
#
# [3310] Remove Methods From Project
#
# @lc code=start
class Solution {
public:
    vector<int> remainingMethods(int n, int k, vector<vector<int>>& invocations) {
        // Build adjacency list
        vector<vector<int>> graph(n);
        for (auto& inv : invocations) {
            graph[inv[0]].push_back(inv[1]);
        }
        
        // Find all suspicious methods (reachable from k)
        vector<bool> suspicious(n, false);
        queue<int> q;
        q.push(k);
        suspicious[k] = true;
        
        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            for (int next : graph[curr]) {
                if (!suspicious[next]) {
                    suspicious[next] = true;
                    q.push(next);
                }
            }
        }
        
        // Check if any non-suspicious method invokes a suspicious method
        bool canRemove = true;
        for (auto& inv : invocations) {
            if (!suspicious[inv[0]] && suspicious[inv[1]]) {
                canRemove = false;
                break;
            }
        }
        
        // Build result
        vector<int> result;
        if (canRemove) {
            // Return only non-suspicious methods
            for (int i = 0; i < n; i++) {
                if (!suspicious[i]) {
                    result.push_back(i);
                }
            }
        } else {
            // Return all methods
            for (int i = 0; i < n; i++) {
                result.push_back(i);
            }
        }
        
        return result;
    }
};
# @lc code=end

File: 3310.remove-methods-from-project.java
#
# @lc app=leetcode id=3310 lang=java
#
# [3310] Remove Methods From Project
#
# @lc code=start
import java.util.*;
class Solution {
    public List<Integer> remainingMethods(int n, int k, int[][] invocations) {
        // Build the invocation graph: who invokes whom
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] inv : invocations) {
            int a = inv[0], b = inv[1];
            graph.get(a).add(b);
        }
        // Find all suspicious methods via BFS from k
        boolean[] suspicious = new boolean[n];
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(k);
        suspicious[k] = true;
        while (!queue.isEmpty()) {
            int curr = queue.poll();
            for (int nei : graph.get(curr)) {
                if (!suspicious[nei]) {
                    suspicious[nei] = true;
                    queue.offer(nei);
                }
            }
        }
        // Check if any method outside suspicious invokes into suspicious
        for (int i = 0; i < n; i++) {
            if (!suspicious[i]) {
                for (int nei : graph.get(i)) {
                    if (suspicious[nei]) {
                        // Not allowed to remove
                        List<Integer> all = new ArrayList<>();
                        for (int j = 0; j < n; j++) all.add(j);
                        return all;
                    }
                }
            }
        }
        // Safe to remove all suspicious methods
        List<Integer> res = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (!suspicious[i]) res.add(i);
        }
        return res;
    }
}
# @lc code=end